lls/arp.c

```diff
@@ -0,0 +1,236 @@
+
+	/* Set-up Ethernet header. */
+	eth_hdr = rte_pktmbuf_mtod(created_pkt, struct ether_hdr *);
+	ether_addr_copy(&iface->eth_addr, &eth_hdr->s_addr);
+	if (ha == NULL)
+		memset(&eth_hdr->d_addr, 0xFF, ETHER_ADDR_LEN);
+	else
+		ether_addr_copy(ha, &eth_hdr->d_addr);
+	eth_hdr->ether_type = rte_cpu_to_be_16(ETHER_TYPE_ARP);
+
+	/* Set-up ARP header. */
+	arp_hdr = (struct arp_hdr *)(eth_hdr + 1);
+	arp_hdr->arp_hrd = rte_cpu_to_be_16(ARP_HRD_ETHER);
+	arp_hdr->arp_pro = rte_cpu_to_be_16(ETHER_TYPE_IPv4);
+	arp_hdr->arp_hln = ETHER_ADDR_LEN;
+	arp_hdr->arp_pln = sizeof(uint32_t);
+	arp_hdr->arp_op = rte_cpu_to_be_16(ARP_OP_REQUEST);
+	ether_addr_copy(&iface->eth_addr, &arp_hdr->arp_data.arp_sha);
+	arp_hdr->arp_data.arp_sip = iface->ip4_addr.s_addr;
+	memset(&arp_hdr->arp_data.arp_tha, 0, ETHER_ADDR_LEN);
+	arp_hdr->arp_data.arp_tip = *(const uint32_t *)ip_be;
+
+	ret = rte_eth_tx_burst(iface->id, tx_queue, &created_pkt, 1);
+	if (ret <= 0) {
+		rte_pktmbuf_free(created_pkt);
+		RTE_LOG(ERR, GATEKEEPER,
+			"lls: could not transmit an ARP request\n");
+	}
+}
+
+/*
+ * A Gratuitous ARP is an ARP request that serves as an announcement of
+ * a neighbor's mapping. The sender and target IP address should be the same,
+ * AND the target Ethernet address should be the same as the sender Ethernet
+ * address OR zero.
+ */
+static inline int
+is_garp_pkt(const struct arp_hdr *arp_hdr)
+{
+	return (arp_hdr->arp_data.arp_sip == arp_hdr->arp_data.arp_tip) &&
+		(is_zero_ether_addr(&arp_hdr->arp_data.arp_tha) ||
+		is_same_ether_addr(&arp_hdr->arp_data.arp_tha,
+			&arp_hdr->arp_data.arp_sha));
+}
+
+int
+process_arp(struct lls_config *lls_conf, struct gatekeeper_if *iface,
+	uint16_t tx_queue, struct rte_mbuf *buf, struct ether_hdr *eth_hdr)
+{
+	struct in_addr target_ip_be;
+	struct arp_hdr *arp_hdr;
+	int is_garp;
+	uint16_t pkt_len = rte_pktmbuf_data_len(buf);
+
+	if (pkt_len < sizeof(*eth_hdr) + sizeof(*arp_hdr)) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: %s interface received ARP packet of size %hu bytes, but it should be at least %zu bytes\n",
+			iface->name, pkt_len,
+			sizeof(*eth_hdr) + sizeof(*arp_hdr));
+		return -1;
+	}
+
+	arp_hdr = rte_pktmbuf_mtod_offset(buf, struct arp_hdr *,
+		sizeof(struct ether_hdr));
+
+	if (unlikely(arp_hdr->arp_hrd != rte_cpu_to_be_16(ARP_HRD_ETHER) ||
+		     arp_hdr->arp_pro != rte_cpu_to_be_16(ETHER_TYPE_IPv4) ||
+		     arp_hdr->arp_hln != ETHER_ADDR_LEN ||
+		     arp_hdr->arp_pln != sizeof(uint32_t)))
+		return -1;
+
+	/* TODO If sip is not in the same subnet as our IP address, drop. */
+
+	/*
+	 * Check if the target IP address is us or if it's
+	 * a Gratuitous ARP. If neither are true, then it
+	 * shouldn't have been sent to us.
+	 */
+	target_ip_be.s_addr = arp_hdr->arp_data.arp_tip;
+	is_garp = is_garp_pkt(arp_hdr);
+	if (!is_garp && (iface->ip4_addr.s_addr != target_ip_be.s_addr))
+		return -1;
+
+	/* Update cache with source resolution, regardless of operation. */
+	lls_cache_update(&lls_conf->arp_cache,
```

Comment:
It's in lls/cache.c. When I view this pull request on GitHub, the review system chooses to not display lls/cache.c -- you have to click "show diff" in order for it to be shown. I don't know why.